### ISSUE TITLE
Migration of big GitHub repo fails

### ISSUE THREAD

**Issue by davidak** [go-gitea/gitea#13241]
- Gitea version (or commit ref): 1.13.0+rc1
- Git version: 2.20.1
- Operating system: debian stable 10.6
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [X] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [X] Yes https://try.gitea.io/nixos/nixpkgs https://try.gitea.io/nixos/nixpkgs/settings/branches
  - [ ] No
- Log gist: https://gist.github.com/Codeberg-org/a4e32102e761c857ecf46e3fb6f22237

## Description

When i try to migrate [nixpkgs](https://github.com/NixOS/nixpkgs) to the [Codeberg test server](https://codeberg-test.org/), it always fails. We have increased the migration timeouts x100.

When we wait or open https://codeberg-test.org/NixOS/nixpkgs/settings/branches, we get a trace with error in `setting_protected_branch.go:34`.

It is not a hardware limitation.

## Screenshots

![Screenshot from 2020-10-21 15-14-34](https://user-images.githubusercontent.com/91113/96740435-21b41d00-13c1-11eb-8693-fd77318ec671.png)

![Screenshot from 2020-10-21 15-23-41](https://user-images.githubusercontent.com/91113/96739172-ccc3d700-13bf-11eb-8c8a-5aa19f68e320.png)

![Screenshot from 2020-10-21 19-03-50](https://user-images.githubusercontent.com/91113/96753996-f46f6b00-13d0-11eb-9f64-ea03f48b19f3.png)
![Screenshot from 2020-10-21 19-08-58](https://user-images.githubusercontent.com/91113/96754004-f6392e80-13d0-11eb-8697-3b1e48c1d150.png)

**Comment by jolheiser** [comment]
For anyone looking in to it, this is the line referenced.

https://github.com/go-gitea/gitea/blob/ba97c0e98bc97957d6fd9bfd3db5768813c58ff3/routers/repo/setting_protected_branch.go#L34

**Comment by davidak** [comment]
@jolheiser are you a bot? :smile: (because of the instant reaction)

Thanks for the information!

**Comment by Codeberg-org** [comment]
To add some more color, this log line (also referenced in the gist) indicates that gh throttled API calls, fetched data may or may not be truncated or incomplete:

```raw
modules/task/task.go:51:handle() [E] Run task failed: GET https://api.github.com/repos/NixOS/nixpkgs: 403 API rate limit of 5000 still exceeded until 2020-10-21 16:33:28 +0200 CEST, not making remote request. [rate reset in 34m49s]
```

**Comment by jolheiser** [comment]
I really hope not! 🤖 

Also, for what it's worth I believe this would benefit from https://github.com/go-gitea/gitea/pull/12244 as well. 

The biggest blocker for moving Gitea to gitea.com is a combination of network and the fact that it needs to make numerous API calls (which may exceed limit).

**Comment by davidak** [comment]
And nixpkgs is way bigger. It has 100,000 issues and pull requests. Probably the second most on GitHub (1. is vscode).

**Comment by Codeberg-org** [comment]
Even if API limits are hit, migration should not fail but get throttled, too? Ideally with some estimate of ETA?

**Comment by jolheiser** [comment]
It seems we do have some code for sleeping if we hit a rate limit.

https://github.com/go-gitea/gitea/blob/f0fe5683feb2799e3ca467bc75fd77871b65452b/modules/migrations/github.go#L113-L128

Do you have any logs surrounding that rate limit error that may indicate where in the migration process it was?

**Comment by davidak** [comment]
>Dumping to disk may not help if we are hitting a rate limit anyways.

We got this trace 3 times without seeing the api limit message, so it is not related to this issue. We got the limit after trying too often and waited.

**Comment by davidak** [comment]
>Even if API limits are hit, migration should not fail but get throttled, too? Ideally with some estimate of ETA?

That's a separate issue.

**Comment by 6543** [comment]
@davidak I dont know how expencive it is, but if one of NixOS maintainers buy a GitHub enterprice account for the periode of migration, it would be faster in genera ... (higher api rate limit)

**Comment by davidak** [comment]
@6543 do you mean GitHub Pro (https://github.com/account/upgrade) or where can you buy it?

![Screenshot from 2020-10-21 19-43-40](https://user-images.githubusercontent.com/91113/96757589-ecfe9080-13d5-11eb-80f6-286dd77bca42.png)

That seem like a useful information that should be part of documentation.

But you should not have to buy your data free in general.

**Comment by jolheiser** [comment]
@davidak He's referring to https://github.com/enterprise which is basically self-hosted GitHub.

**Comment by 6543** [comment]
@davidak also as suggested on <https://stackoverflow.com/questions/16732103/increase-github-api-limit> just contact github directly?

**Comment by davidak** [comment]
Combine your tokens to free the code :fist_raised: 

Might not be a good idea to give someone else your token...

Asking github would be an option, but it would be best if it just don't fail and respect the API limit.

But this issue here has to be fixed first before we can try again.

**Comment by Codeberg-org** [comment]
> 
> Do you have any logs surrounding that rate limit error that may indicate where in the migration process it was?

Could not see anything in the log; is some special app.ini config needed to enable this logging?

**Comment by lunny** [comment]
If you are migrating multiple projects at the same time, there are some chances to encounter the rate limit. The rate limit is per github acccount I think.

**Comment by TheFrenchGhosty** [comment]
I can confirm this issue. This is a MASSIVE problem.

From what I discovered Codeberg doesn't seem to be affected by this, but every instance where I tried (including my own) is affected.

@Codeberg-org  do you guys have any specific settings for migrations? I couldn't find anything special in this config file https://codeberg.org/Codeberg/build-deploy-gitea/src/branch/master/etc/gitea/conf/app.ini

**Comment by billewanick** [comment]
I'm also running into this issue. I'm trying to create a pull mirror of Nixpkgs on my own Gitea instance.
The server url is [git.ewanick.com](https://git.ewanick.com/bill). The server configuration is defined [here](https://git.ewanick.com/bill/deployed-nixos-server-and-apps/src/branch/main/server-config/configuration.nix), and the gitea instance is defined [in this module](https://git.ewanick.com/bill/deployed-nixos-server-and-apps/src/branch/main/nixos-apps/git.ewanick.com.nix).

The details from the Gitea Admin console are:
```
Gitea version (or commit ref): 1.19.3 built with go1.20.4 : pam, sqlite, sqlite_unlock_notify
Git version: 2.40.1, Wire Protocol Version 2 Enabled
Operating system: NixOS 23.05 "Stoat"
Database: SQLite
```

The server is running on a 2 Core / 4GB Ram Linode dedicated server.
I am able to manually clone the Nixpkgs repo to the gitea `repositories` folder and recover it as an abandoned repo, but then it won't let me set up the pull mirror, only push.

I have updated [the timeouts to be extra long](https://git.ewanick.com/bill/deployed-nixos-server-and-apps/src/branch/main/nixos-apps/git.ewanick.com.nix#L15-L20) and it still fails:
```
Migration Timeout - 6000 seconds
Mirror Update Timeout - 6000 seconds
Clone Operation Timeout - 300 seconds
Pull Operation Timeout - 6000 seconds
GC Operation Timeout - 6000 seconds
```

I am using a GitHub Personal Access Token to try and get around rate-limiting.

Any help with this would be greatly appreciated, including helping me figure out where the logs are.

**Comment by 6543** [comment]
if it's only about the git data - dont use github but just the git option without your token and clone via https

and you can just add `00` to all your timeouts temporary - that should do it

**Comment by billewanick** [comment]
Changed all my timeouts to 60000, and tried the git option with https.
Still failed when I tried it several times.
In the Firefox console there was one error, a 404 on a GET request. `task `12` does not exist`

**Comment by bircni** [comment]
I migrated 3 large large repos today no problem in the 4th i got rate limited
